src/pip/_internal/index/package_finder.py

```diff
@@ -608,6 +607,7 @@ def __init__(
 
         # These are boring links that have already been logged somehow.
         self._logged_links = set()  # type: Set[Link]
+        self._logged_links_rp = set()  # type: Set[str]
```

Comment:
The name reads very wrong to me. This does not collect links, and `rp` is not a common acronym for anything.

I wonder if there’s a way to collect these with `_logged_links` as well. Both are just skipped links but for different reasons, so it feels more right to me if we could combine the two.

Reply:
I thought that `_logged_links_requires_python` is a tad too long and i saw `rp` in `models/link.py` though that is just a variable name.

I don't recall finding a way to distinguish links skipped because it requires non-supported Python version if we group them together. There might be links for which `link.requires_python` is true, but skipped for other reasons?

Reply:
also, initially I tried using `Set[Link]` to match `_logged_links` type, but getting the dependency version from a given link is a very long step, hence using the already found version in `evaluate_link`.

Reply:
> There might be links for which `link.requires_python` is true, but skipped for other reasons?

There should be—for example, a wheel for another platform can have Requires-Python set, but should be skipped because it’s not for this target machine.

Maybe we can make `_logged_links` a `Set[Link, Reason]` instead? This way we can check the second argument to get a list of links skipped due to `Requires-Python` mismatches.

Reply:
> Maybe we can make _logged_links a Set[Link, Reason] instead? This way we can check the second argument to get a list of links skipped due to Requires-Python mismatches.

do you mean `Set[Tuple[Link, str]]` where `str` is the reason?

I tried implementing the changes [here](https://github.com/ronaudinho/pip/commit/7584278e2162a5d9a91e53cba36ce71614244e23) to avoid triggering actions for small changes. The printed warning shows duplicate versions (since they come from different links after all), I am not sure if [this](https://github.com/ronaudinho/pip/commit/459b0d897e3d8a9c66dc20fa5b9af328b6a3404b) is necessary. Let me know if they look better, I'll push it here.

Reply:
Yes, version deduplication is necessary. The *Could not find a version that satisfies the requirement* message also contains similar logic (see `_format_versions` in `PackageFinder.find_requirement()`).

> do you mean `Set[Tuple[Link, str]]` where `str` is the reason?

Yes.